transformer/plugins/contracts.py

```diff
@@ -52,109 +47,131 @@
-# deal with a full, immutable list of tasks in plugins that independently
-# operate on each task implies a lot of verbosity and redundancy.
-# For these reasons, other plugin kinds were created to offer a more varied
-# choice for plugin implementers.
-# See https://github.com/zalando-incubator/Transformer/issues/10.
-OnTaskSequence = Callable[[Sequence[Task]], Sequence[Task]]
+def contract(f: Plugin) -> Contract:
+    """
+    Returns the contract associated to a plugin function.
 
+    :raise InvalidPluginError: if f is not a plugin.
+    """
+    try:
+        return getattr(f, "_transformer_plugin_contract")
+    except AttributeError:
+        raise InvalidPluginError(f) from None
 
-Plugin = Union[OnTask, OnScenario, OnPythonProgram, OnTaskSequence]
 
-for plugin_type in Plugin.__args__:
-    _register_contract(plugin_type)
+_T = TypeVar("_T")
 
 
-def isvalid(plugin_type: type, obj: Any) -> Decision:
+def apply(plugins: Iterable[Plugin], init: _T) -> _T:
     """
-    Checks whether obj is an implementation of the plugin contract plugin_type.
-    The return value is basically a boolean, with an additional string
-    describing the reason for this decision.
-
-    :param plugin_type: plugin contract to verify obj against
-    :param obj: any Python object
-    :return: whether obj is conform to the plugin_type contract
-    :raise TypeError: if plugin_type is not a plugin contract
+    Applies each plugin to init in order, and returns the result.
+
+    This just wraps a very simple but common operation.
     """
-    if plugin_type is Plugin:
-        return Decision.any(
-            (isvalid(t, obj) for t in Plugin.__args__),
-            "should be valid for a Plugin subtype",
-        )
+    for p in plugins:
+        init = p(init)
+    return init
 
-    if not callable(obj):
-        return Decision.no(f"{obj!r} is not a function")
 
-    try:
-        validator = _PLUGIN_VALIDATORS[plugin_type]
-    except KeyError:
-        raise TypeError(f"no Plugin contract registered for {plugin_type}")
+_BASE_CONTRACTS = (
+    Contract.OnTask,
+    Contract.OnTaskSequence,
+    Contract.OnScenario,
+    Contract.OnPythonProgram,
+)
 
-    try:
-        actual_signature = inspect.signature(obj)
-    except (ValueError, TypeError) as err:
-        return Decision.no(f"could not extract signature from {obj!r}: {err}")
 
-    return Decision.whether(
-        validator(actual_signature), f"{obj.__name__!r} should implement {plugin_type}"
-    )
+def group_by_contract(plugins: Iterable[Plugin]) -> DefaultDict[Contract, List[Plugin]]:
+    """
+    Groups plugins in lists according to their contracts.
+    Each plugin is found in as many lists as it implements base contracts.
+    Lists keep the order of the original plugins iterable.
+    """
+    res = defaultdict(list)
+    for p in plugins:
+        c = contract(p)
+        for bc in _BASE_CONTRACTS:
+            if c & bc:
```

Comment:
What is the purpose of the bitwise operation here?

Reply:
Contract is an `enum.Flag`, which means its members can be combined with bitwise operators, by analogy with how "flags" are generally implemented in C (power-of-two integers).

This is useful for contracts because Transformer just defines the basic cases (like `OnTask` and `OnScenario`) and users can combine them if they need (like `OnTask | OnScenario`, which defines a plugin that works with both Task and Scenario objects as input).

That particular bitwise operation `c & bc`, where `c` is an arbitrary contract and `bc` is a "base contract", returns (in boolean context, like here in `if`) whether `c` and `bc` have contracts in common.

You can think of it as if contracts were implemented as power-of-two integers or as sets, and you are using the corresponding `&` operator that tells you whether there is an intersection between these two sets of contracts.

Does that make it clearer?

Reply:
Thanks! The bit I was missing to understand this is the `enum.Flag` usage 😄 Do you think you could add a very short comment explaining it (without going to details)?

Reply:
Yes, would you rather see this comment close to the `&` or close to the enum definition?

Reply:
I think it make sense to add it here, but I trust you make the best choice